File: content/snippets/scala3/interview_lcci_08_06_hanota.scala
import scala.collection.mutable.ArrayBuffer

object Lcci0806Hanota:
    def hanota(A: ArrayBuffer[Int], B: ArrayBuffer[Int], C: ArrayBuffer[Int]): Unit =
        def move(n: Int, src: ArrayBuffer[Int], aux: ArrayBuffer[Int], dst: ArrayBuffer[Int]): Unit =
            if (n == 0) return
            move(n - 1, src, dst, aux)
            dst += src.remove(src.length - 1)
            move(n - 1, aux, src, dst)
        move(A.length, A, B, C)

    @main def main(): Unit =
        val A = ArrayBuffer(2, 1, 0)
        val B = ArrayBuffer[Int]()
        val C = ArrayBuffer[Int]()
        hanota(A, B, C)
        println(s"A: ${A.mkString("[", ",", "]")}")
        println(s"B: ${B.mkString("[", ",", "]")}")
        println(s"C: ${C.mkString("[", ",", "]")}")

File: content/snippets/cpp/interview_lcci_08_06_hanota.cpp
#include <iostream>
#include <vector>
#include <functional>
using namespace std;

void hanota(vector<int>& A, vector<int>& B, vector<int>& C) {
    function<void(int, vector<int>&, vector<int>&, vector<int>&)> move =
        [&](int n, vector<int>& src, vector<int>& aux, vector<int>& dst) {
            if (n == 0) return;
            move(n - 1, src, dst, aux);
            dst.push_back(src.back());
            src.pop_back();
            move(n - 1, aux, src, dst);
        };
    move((int)A.size(), A, B, C);
}

int main() {
    vector<int> A = {2, 1, 0}, B = {}, C = {};
    hanota(A, B, C);
    cout << "A: ["; for (int i = 0; i < (int)A.size(); i++) cout << (i?",":"") << A[i]; cout << "]" << endl;
    cout << "B: ["; for (int i = 0; i < (int)B.size(); i++) cout << (i?",":"") << B[i]; cout << "]" << endl;
    cout << "C: ["; for (int i = 0; i < (int)C.size(); i++) cout << (i?",":"") << C[i]; cout << "]" << endl;
    return 0;
}
